### PR TITLE
*: add foreign key cascade delete when execute delete statement

### LINKED ISSUE #38376: support foreign key cascade delete
## Dev Task

support foreign key cascade delete behaviour.

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #38376

Problem Summary: add foreign key cascade delete when execute delete statement

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by ekexium — COMMENTED** [review]
It seems `explain analyze` doesn't work with cascade delete, e.g. `explain analyze delete from` the parent table doesn't delete the referencing row in the child table.

File: errno/errname.go [hunk 1]
```diff
@@ -838,6 +838,7 @@ var MySQLErrName = map[uint16]*mysql.ErrMessage{
 	ErrAccountHasBeenLocked:                                  mysql.Message("Access denied for user '%s'@'%s'. Account is locked.", nil),
 	ErrWarnConflictingHint:                                   mysql.Message("Hint %s is ignored as conflicting/duplicated.", nil),
 	ErrUnresolvedHintName:                                    mysql.Message("Unresolved name '%s' for %s hint", nil),
+	ErrForeignKeyCascadeDepthExceeded:                        mysql.Message("Foreign key cascade delete/update exceeds max depth of 15.", nil),
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply errno/errname.go:841]: done.

**Comment by crazycs520** [comment]
> It seems `explain analyze` doesn't work with cascade delete, e.g. `explain analyze delete from` the parent table doesn't delete the referencing row in the child table.



> It seems `explain analyze` doesn't work with cascade delete, e.g. `explain analyze delete from` the parent table doesn't delete the referencing row in the child table.

Nice catch, since the PR already too large, I create an issue https://github.com/pingcap/tidb/issues/38513 for it and fix it in another PR.

**Review by you06 — APPROVED** [review]
This PR LGTM, and I'm still curious about whether it's ok to do a recursive FK handling for update statements.

File: executor/adapter.go [hunk 2]
```diff
@@ -561,18 +565,121 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
...
+
+func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool, depth int) error {
 	exec, ok := e.(WithForeignKeyTrigger)
 	if !ok {
 		return nil
 	}
+	originStmtCtx := a.Ctx.GetSessionVars().StmtCtx
+	defer func() {
+		// Reset to the original stmtCtx.
+		a.Ctx.GetSessionVars().StmtCtx = originStmtCtx
```
Comment (ekexium) [inline executor/adapter.go:600]: I think `stmtctx` is not designed to be used like this. It's way too complex to reason every field in it in this case. I bet there are bugs here.

**Review by ekexium — APPROVED** [review]
It looks good as an experimental work, but I'd say we need far more tests, solid reasoning and probably some refactoring to make it production-ready.

They key here I think is in some places the delete statement and the casdading statement(s) are treated as separate/individual statements (e.g. they have individual statement contexts), while in some places they may be treated as one (they may have shared states like `ctx`?). The inconsistency in the design is bug-prone.

It's very likely there are compatibility issues with other features. For example, it seems tracing doesn't work well with cascade statements, not a big problem though.

File: executor/adapter.go [hunk 3]
```diff
@@ -561,18 +565,121 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
...
+		return err
+	}
+	if err := e.Open(ctx); err != nil {
+		terror.Call(e.Close)
+		return err
+	}
+	_, _, e, err = a.handleNoDelay(ctx, e, isPessimistic)
+	if err != nil {
+		return err
+	}
```
Comment (cfzjywxk) [inline executor/adapter.go:637]: I'm not quite clear about the previous state assumptions, maybe it's better to add some descriptions to explain the safety, which states of a statement execution may change or may not change. 
For example, should the executions refresh the `for_update_ts` or they should always use the same one?

File: executor/foreign_key.go [hunk 4]
```diff
@@ -503,3 +519,164 @@ func (fkc FKCheckExec) checkRows(ctx context.Context, sc *stmtctx.StatementConte
...
+	sctx := fkc.b.ctx
+	exec, ok := sctx.(sqlexec.RestrictedSQLExecutor)
+	if !ok {
+		return nil, nil
+	}
+	stmtNode, err := exec.ParseWithParams(ctx, sqlStr)
+	if err != nil {
+		return nil, err
+	}
+	err = ResetContextOfStmt(sctx, stmtNode)
```
Comment (cfzjywxk) [inline executor/foreign_key.go:611]: The `ResetContextOfStmt` is called by the execution framework. The function `uildFKCascadePlan` does something the same as the framework does, maybe it's better to reuse the existing functions by making necessary changes to them.

File: executor/adapter.go (hunk 2 above)
Comment (crazycs520) [reply executor/adapter.go:602]: You are right, I should handle `stmtctx` carefully.

File: executor/foreign_key.go (hunk 4 above)
Comment (crazycs520) [reply executor/foreign_key.go]: It really doesn't quite fit here, so it has been deleted.

File: executor/adapter.go [hunk 5]
```diff
@@ -561,18 +565,121 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
...
+}
+
+var maxForeignKeyCascadeDepth = 15
+
+func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool, depth int) error {
 	exec, ok := e.(WithForeignKeyTrigger)
 	if !ok {
 		return nil
 	}
+	originStmtCtx := a.Ctx.GetSessionVars().StmtCtx
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/adapter.go]: already remove the reset logic here.

File: executor/adapter.go (hunk 3 above)
Comment (crazycs520) [reply executor/adapter.go:636]: Great, maybe I should reuse `handleNoDelay` here, write an `handleFKExecutor` method to handle it maybe better.

**Comment by crazycs520** [comment]
> 

It's fixed now.

File: executor/adapter.go (hunk 2 above)
Comment (crazycs520) [reply executor/adapter.go]: I have refactored the logic here, PTAL

File: executor/adapter.go (hunk 3 above)
Comment (crazycs520) [reply executor/adapter.go]: I have refactored the logic here, PTAL

File: executor/foreign_key.go (hunk 4 above)
Comment (crazycs520) [reply executor/foreign_key.go]: I have refactored the logic here, PTAL

**Comment by crazycs520** [comment]
> > > It seems `explain analyze` doesn't work with cascade delete, e.g. `explain analyze delete from` the parent table doesn't delete the referencing row in the child table.
> > 
> > 
> > > It seems `explain analyze` doesn't work with cascade delete, e.g. `explain analyze delete from` the parent table doesn't delete the referencing row in the child table.
> > 
> > 
> > Nice catch, since the PR already too large, I create an issue #38513 for it and fix it in another PR.
> 
> I mean it doesn't do the cascade delete in an `explain analyze`, not only missing information in returned results.

It's fixed now.

File: executor/adapter.go (hunk 2 above)
Comment (ekexium) [reply executor/adapter.go]: Does it mean 2 statements may share some of the same `stmtctx` now? I doubt that's still error-prone

Comment (crazycs520) [reply executor/adapter.go]: yes..., but the foreign key cascade SQL shouldn't be treated as a statement, it is another  `Executor`.

Comment (crazycs520) [reply executor/adapter.go]: I'm importing more tests from MySQL to test the correctness of this.
https://github.com/pingcap/tidb-test/pull/1993

File: executor/foreign_key.go [hunk 6]
```diff
@@ -503,3 +521,154 @@ func (fkc FKCheckExec) checkRows(ctx context.Context, sc *stmtctx.StatementConte
...
+	if len(fkc.fkValues) == 0 {
+		return nil, nil
+	}
+	var sqlStr string
+	var err error
+	switch fkc.tp {
+	case plannercore.FKCascadeOnDelete:
+		sqlStr, err = GenCascadeDeleteSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, fkc.fk, fkc.fkValues)
+	}
+	if err != nil || sqlStr == "" {
```
Comment (you06) [inline executor/foreign_key.go:600]: If `err == nil && sqlStr == ""`, no error will be returned and the statement is executed successfully, though this is unreachable, it's better to return error in this case.

File: executor/adapter.go [hunk 7]
```diff
@@ -561,22 +556,124 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
...
+	if sc.ForeignKeyTriggerCtx.SavepointName == "" {
+		return nil
+	}
+	txn, err := a.Ctx.Txn(false)
+	if err != nil || !txn.Valid() {
+		return err
+	}
+	savepointRecord := a.Ctx.GetSessionVars().TxnCtx.RollbackToSavepoint(sc.ForeignKeyTriggerCtx.SavepointName)
+	if savepointRecord == nil {
+		return errors.Errorf("foreign key cascade savepoint '%s' not found, should never happen", sc.ForeignKeyTriggerCtx.SavepointName)
```
Comment (you06) [inline executor/adapter.go:669]: Since the correctness may be broken, abort the transaction in this case?

File: executor/foreign_key.go (hunk 6 above)
Comment (crazycs520) [reply executor/foreign_key.go]: great, done.

File: executor/adapter.go (hunk 7 above)
Comment (crazycs520) [reply executor/adapter.go]: great, done.

**Review by you06 — APPROVED** [review]

**Review by ekexium — APPROVED** [review]

**Pull request merged**